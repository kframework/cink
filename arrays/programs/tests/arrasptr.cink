/*

  This example exhibits the relationship between
  arrays and pointers.

*/

int main() {
  int a[5], i;
  int s = 0, *p;
  i = 0;
  while (i < 5) {
    a[i] = 2*i;
    ++i;
  }
  p = & (a[0]);
  while (p <= &(a[4])) {
    s = s + *(p++);
  }
  cout << s << endl;
  return 0;
}

  
