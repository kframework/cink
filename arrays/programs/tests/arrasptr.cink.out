20
<T>
    <in>
        ListItem(#buffer ( "" ))
        ListItem(#istream ( 0 ))
    </in>
    <out>
        ListItem(#ostream ( 1 ))
        ListItem(#buffer ( "" ))
    </out>
    <control>
        <k>
            .K
        </k>
        <return>
            0
        </return>
        <fstack>
            .List
        </fstack>
        <sideEffects>
            .Bag
        </sideEffects>
        <env>
            main |-> #symScalLoc(1)
        </env>
    </control>
    <genv>
        main |-> #symScalLoc(1)
    </genv>
    <store>
        seq-array ( 5 , #symLoc(2) ) |-> store ( (store ( (store ( (store 
          ( (store ( (const-array ( 5 , 0 )) , 0 , 0 )) , 1 , 2 )) , 2 , 4 )) 
          , 3 , 6 )) , 4 , 8 )
        #symScalLoc(1) |-> lambda ( .Decls @ ((((((((((int (a [ 5 ])) ;) ((int
           i) ;)) ((((int s) ;) ((s = 0) ;)) (p of (pointer to int) ;))) ((i =
           0) ;)) (while ( (i < 5) ) ({ ((((a [ i ]) = (2 * i)) ;) ((++ i) ;))
           }))) ((p = (& (a [ 0 ]))) ;)) (while ( (p <= (& (a [ 4 ]))) ) ({ ((
          s = (s + (* (p ++)))) ;) }))) (((cout << s) << endl) ;)) (return 0 ;
          )) )
        #symScalLoc(3) |-> 5
        #symScalLoc(4) |-> 20
        #symScalLoc(5) |-> select ( (seq-array ( 5 , #symLoc(2) )) , 5 )
    </store>
</T>
